RHDEVDOCS-3386 - How to configure Prometheus Adapter audit log levels (#44229)
openshift/openshift-docs: modules/monitoring-setting-audit-log-levels-for-the-prometheus-adapter.adoc

[role=_abstract]
In default platform monitoring, you can configure the audit log level for the Prometheus Adapter.

.Prerequisites

* You have installed the OpenShift CLI (`oc`).
* You have access to the cluster as a user with the `cluster-admin` role.
* You have created the `cluster-monitoring-config` `ConfigMap` object.

.Procedure

You can set an audit log level for the Prometheus Adapter in the default `openshift-monitoring` project:

. Edit the `cluster-monitoring-config` `ConfigMap` object in the `openshift-monitoring` project:
+
[source,terminal]
----
$ oc -n openshift-monitoring edit configmap cluster-monitoring-config
----

. Add `profile:` under `audit` in the `k8sPrometheusAdapter` section under `data/config.yaml`:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    k8sPrometheusAdapter:
      audit:
        profile: <audit_log_level> <1>
----
<1> The audit log level to apply to the Prometheus Adapter. 

. Set the audit log level by using one of the following values for the `profile:` parameter:
+
* `None`: Do not log events.

> * `Metadata`: Log only the metadata for the request, such as user, timestamp, and so forth. The request text and the response text will not be logged.

Reviewer: `Metadata` is the default audit level, we'd better mention it

Author: Added a sentence to the description indicating that `Metadata` is the default value.

> * `Request`: - Log only the metadata and the request text but not the response text. This option does not apply for non-resource requests.

Reviewer: `- Log` should be ` Log`

Author: fixed

> * `RequestResponse`: - Log event metadata, request text, and response text. This option does not apply for non-resource requests.

Reviewer: `- Log` should be ` Log`

Author: fixed

. Save the file to apply the changes. The pods for the Prometheus Adapter restart automatically when you apply the change.
+
[WARNING]
====
When changes are saved to a monitoring config map, the pods and other resources in the related project might be redeployed. The running monitoring processes in that project might also be restarted.
====

.Verification

> . Check that the pods for the Prometheus Adapter are running. The following example lists the status of pods in the `openshift-monitoring` project:

Reviewer: the example set `profile: request`, we should mention it

Reviewer: and we need to check the prometheus-adapter file, then we know the audit log file path
```
# oc -n openshift-monitoring get deploy prometheus-adapter -o yaml
...
        - --audit-policy-file=/etc/audit/request-profile.yaml
        - --audit-log-path=/var/log/adapter/audit.log
```

Author: added

+
[source,terminal]
----
$ oc -n openshift-monitoring get pods
----

> . Confirm that the correct log level has been applied in the `prometheus-adapter` deployment in the `openshift-monitoring` project. In the following example command, substitute `prometheus-adapter-84484ff64-x99wz` with the pod name for the Prometheus Adapter in your deployment:

Reviewer: only `Confirm that the correct log level has been applied in the `prometheus-adapter` deployment in the `openshift-monitoring` project.` is fine

Reviewer: and we should mention how to check the 
# oc -n openshift-monitoring get deploy prometheus-adapter -oyaml
...
        - --audit-policy-file=/etc/audit/request-profile.yaml
        - --audit-log-path=/var/log/adapter/audit.log

Reviewer: we'd better mention how tho check the audit log level and audit log file path
```
# oc -n openshift-monitoring get deploy prometheus-adapter -oyaml
...
        - --audit-policy-file=/etc/audit/request-profile.yaml
        - --audit-log-path=/var/log/adapter/audit.log
```

Author: - Fixed. Added a new step to check audit log level and audit log  file path

+
[source,terminal]
----

> $ c -n openshift-monitoring exec -c prometheus-adapter prometheus-adapter-84484ff64-x99wz -- cat /etc/audit/request-profile.yaml

Reviewer: `oc -n openshift-monitoring exec -c prometheus-adapter prometheus-adapter-84484ff64-x99wz -- cat /etc/audit/request-profile.yaml`

Author: Changed to `oc -n openshift-monitoring exec -c prometheus-adapter <prometheus_adapter_pod_name> -- cat /etc/audit/request-profile.yaml` to follow OCP doc guidelines for including variables in commands

----
+
.Example output
[source,terminal]
----
"apiVersion": "audit.k8s.io/v1"
"kind": "Policy"
"metadata":
  "name": "Request"
"omitStages":
- "RequestReceived"
"rules":
- "level": "Request"
----
+
[IMPORTANT]
====
If an unrecognized `profile` value is used for the Prometheus Adapter in the `ConfigMap` object, the Prometheus Adapter pods might not restart successfully.
====

> . Review the audit log for the Prometheus Adapter. In the following example command, substitute `prometheus-adapter-84484ff64-x99wz` with the pod name for the Prometheus Adapter in your deployment:

Reviewer: Review the audit log for the Prometheus Adapter.
this is enough

Author: Fixed

+
[source,terminal]
----

> oc -n openshift-monitoring exec -c prometheus-adapter prometheus-adapter-84484ff64-x99wz -- cat /var/log/adapter/audit.log

Reviewer: change to
`oc -n openshift-monitoring exec -c prometheus-adapter ${prometheus-adapter-pod} -- cat /etcvar/log/adapter/audit.log`

Author: Changed to `oc -n openshift-monitoring exec -c prometheus-adapter <prometheus_adapter_pod_name> -- cat /var/log/adapter/audit.log` to follow the OCP docs standard for referring to variables in code and commands.

// Module included in the following assemblies:
//
// * monitoring/configuring-the-monitoring-stack.adoc

:_content-type: PROCEDURE
[id="setting-audit-log-levels-for-the-prometheus-adapter_{context}"]
= Setting audit log levels for the Prometheus Adapter

[role=_abstract]

> In default platform monitoring, you can configure the audit log level for the Prometheus Adapter.

Reviewer 2: When I checked, the capitalization of PA online varies from one source to another. 
Do the following suggestions seem right to you?
- If PA is a proper noun, one drops the definite article, like this:  "...for Prometheus Adapter"? 
- Otherwise, for an ordinary noun, one keeps the definite article and drops the capitalization of adapter, like this: "for the Prometheus adapter"

Author: Thanks, good catch. It should be Prometheus Adapter, so without the article. I'll update it.

.Prerequisites

* You have installed the OpenShift CLI (`oc`).
* You have access to the cluster as a user with the `cluster-admin` role.
* You have created the `cluster-monitoring-config` `ConfigMap` object.

.Procedure

You can set an audit log level for the Prometheus Adapter in the default `openshift-monitoring` project:

. Edit the `cluster-monitoring-config` `ConfigMap` object in the `openshift-monitoring` project:
+
[source,terminal]
----
$ oc -n openshift-monitoring edit configmap cluster-monitoring-config
----

> . Add `profile:` under `audit:` for the `k8sPrometheusAdapter` component under `data/config.yaml`:

Reviewer 2: "...under...under..." 
Reading this, I got started trying to figure out what went under what until I looked at the example.
Would `k8sPrometheusAdapter/audit/profile`-style notation work here?

Author: Change to `Add `profile:` in the `k8sPrometheusAdapter/audit` section under `data/config.yaml``

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    k8sPrometheusAdapter:
      audit:
        profile: <audit_log_level> <1>
----
<1> The audit log level to apply to the Prometheus Adapter. 

. Set the audit log level by using one of the following values for the `profile:` parameter:
+
* `None`: Do not log events.

> * `Metadata`: Log only the metadata for the request, such as user, timestamp, and so forth. The request text and the response text will not be logged. `Metadata` is the default audit log level.

Reviewer 2: Avoid future tense
```suggestion
* `Metadata`: Log only the metadata for the request, such as user, timestamp, and so forth. Do not log the request and response text. `Metadata` is the default audit log level.
```

Reviewer 2: Or how about:
```suggestion
* `Metadata`: Log only the metadata for the request, such as user, timestamp, but not the request and response text. `Metadata` is the default audit log level.
```

Author: fixed

* `Request`: Log only the metadata and the request text but not the response text. This option does not apply for non-resource requests.
* `RequestResponse`: Log event metadata, request text, and response text. This option does not apply for non-resource requests.

. Save the file to apply the changes. The pods for the Prometheus Adapter restart automatically when you apply the change.
+
[WARNING]
====

> When changes are saved to a monitoring config map, the pods and other resources in the related project might be redeployed. The running monitoring processes in that project might also be restarted.

Reviewer 3: I am not sure if this would be case when you only modify the audit setting for Prometheus Adapter. 
However, I am okay with keeping this warning.

Author: ok.

====

.Verification

. In the config map, under `k8sPrometheusAdapter/audit/profile`, set the log level to `Request` and save the file.

. Confirm that the pods for the Prometheus Adapter are running. The following example lists the status of pods in the `openshift-monitoring` project:
+
[source,terminal]
----
$ oc -n openshift-monitoring get pods
----

. Confirm that the audit log level and audit log file path are correctly configured:
+
[source,terminal]
----
$ oc -n openshift-monitoring get deploy prometheus-adapter -o yaml
----
+
.Example output
[source,terminal]
----
...
  - --audit-policy-file=/etc/audit/request-profile.yaml
  - --audit-log-path=/var/log/adapter/audit.log
----

. Confirm that the correct log level has been applied in the `prometheus-adapter` deployment in the `openshift-monitoring` project:
+
[source,terminal]
----

> $ oc -n openshift-monitoring exec -c prometheus-adapter <prometheus_adapter_pod_name> -- cat /etc/audit/request-profile.yaml

Reviewer 3: we can use the deployment itself to confirm this by 
```
oc -n openshift-monitoring exec deploy/prometheus-adapter -c prometheus-adapter -- cat /etc/audit/request-profile.yaml
```

----
+
.Example output
[source,terminal]
----
"apiVersion": "audit.k8s.io/v1"
"kind": "Policy"
"metadata":
  "name": "Request"
"omitStages":
- "RequestReceived"
"rules":
- "level": "Request"
----
+
[IMPORTANT]
====

> If an unrecognized `profile` value is used for the Prometheus Adapter in the `ConfigMap` object, the Prometheus Adapter pods might not restart successfully.

Reviewer 3: The current behaviour which we may want to document is that ...

> If the profile is invalid, no changes will be made to prometheus adapter (if one exists) and an error will be logged by cluster-monitoring-operator.  

cc: @simonpasquier  Any thoughts?

Reviewer: yes, cmo would ignore the invalid value and stay as its previous setting

Author: Switched the admonition from IMPORTANT to NOTE and updated the text to read `If an unrecognized `profile` value is entered for the Prometheus Adapter in the `ConfigMap` object, no changes are made to Prometheus Adapter (if one exists), and an error is logged by the Cluster Monitoring Operator.`

> If you enter an unrecognized `profile` value for the Prometheus Adapter in the `ConfigMap` object, no changes are made to the Prometheus Adapter (if one exists), and an error is logged by the Cluster Monitoring Operator.

Reviewer 2: IBM SG: "...avoid using parentheses in general text."
How about:
```suggestion
If you enter an unrecognized `profile` value in the `ConfigMap` object, no changes are made to the Prometheus Adapter, if one exists, and an error is logged by the Cluster Monitoring Operator.
```

Author: Good catch. Fixed.

====

. Review the audit log for the Prometheus Adapter:
+
[source,terminal]
----

> oc -n openshift-monitoring exec -c <prometheus_adapter_pod_name> -- cat /var/log/adapter/audit.log

Reviewer: missed `$`

Reviewer 3: you can use the `deploy/prometheus-adapter` albeit, if you are interested in a particular pod logs, then this is the right command to execute.

Author: Added the missing `$`. Leaving the rest as-is.

> $ oc -n openshift-monitoring exec -c <prometheus_adapter_pod_name> -- cat /var/log/adapter/audit.log

Reviewer: `$ oc -n openshift-monitoring exec deploy/prometheus-adapter -c prometheus-adapter -- cat /var/log/adapter/audit.log`
use this for consistency, since we used
`$ oc -n openshift-monitoring exec deploy/prometheus-adapter -c prometheus-adapter -- cat /etc/audit/request-profile.yaml`
